integration/flink/shared/src/main/java/io/openlineage/flink/utils/ProtobufUtils.java

```diff
@@ -0,0 +1,164 @@
+        .orElse(false);
+  }
+
+  public static Optional<OpenLineage.SchemaDatasetFacet> convert(
+      OpenLineage openLineage, SerializationSchema serializationSchema) {
+    Optional<Class> protobufSerializedClass = getProtobufSerializeClass(serializationSchema);
+
+    if (protobufSerializedClass.isEmpty()) {
+      return Optional.empty();
+    }
+
+    List<SchemaDatasetFacetFields> fields =
+        getSchemaDatasetFacetFields(openLineage, protobufSerializedClass.get());
+    if (!fields.isEmpty()) {
+      OpenLineage.SchemaDatasetFacetBuilder builder = openLineage.newSchemaDatasetFacetBuilder();
+      return Optional.of(builder.fields(fields).build());
+    }
+
+    return Optional.empty();
+  }
+
+  public static Optional<OpenLineage.SchemaDatasetFacet> convert(
+      OpenLineage openLineage, DeserializationSchema deserializationSchema) {
+    Optional<Class> protobufSerializedClass = getProtobufDeserializeClass(deserializationSchema);
+
+    if (protobufSerializedClass.isEmpty()) {
+      return Optional.empty();
+    }
+
+    List<SchemaDatasetFacetFields> fields =
+        getSchemaDatasetFacetFields(openLineage, protobufSerializedClass.get());
+    if (!fields.isEmpty()) {
+      OpenLineage.SchemaDatasetFacetBuilder builder = openLineage.newSchemaDatasetFacetBuilder();
+      return Optional.of(builder.fields(fields).build());
+    }
+
+    return Optional.empty();
+  }
+
+  static List<SchemaDatasetFacetFields> getSchemaDatasetFacetFields(
+      OpenLineage openLineage, Class protobufSerializedClass) {
+    try {
+      Class<?> staticDefintionClass =
+          Class.forName(protobufSerializedClass.getCanonicalName() + "OuterClass");
+
+      Field descriptorField = FieldUtils.getField(staticDefintionClass, "descriptor", true);
+      FileDescriptor fileDescriptor = (FileDescriptor) descriptorField.get(staticDefintionClass);
+
+      ProtobufFieldResolver fieldResolver = new ProtobufFieldResolver(openLineage);
+      return fileDescriptor.getMessageTypes().stream()
+          .flatMap(d -> d.getFields().stream())
+          .filter(d -> protobufSerializedClass.getName().contains(d.getContainingType().getName()))
+          .map(f -> fieldResolver.resolveField(f))
+          .collect(Collectors.toList());
+    } catch (ClassNotFoundException e) {
+      // swallow it
+      log.warn("Couldn't find OuterClass for {}: {}", protobufSerializedClass, e);
+      return Collections.emptyList();
+    } catch (IllegalAccessException e) {
+      // swallow it
+      log.warn("Couldn't find descriptor property in OuterClass", e);
+      return Collections.emptyList();
+    }
+  }
+
+  private static Optional<Class> getProtobufSerializeClass(
+      SerializationSchema serializationSchema) {
+    Optional<Class> parameterType =
+        Arrays.stream(serializationSchema.getClass().getMethods())
+            .filter(m -> m.getName().equalsIgnoreCase("serialize"))
+            .filter(m -> m.getParameterTypes()[0] != Object.class)
```

Comment:
I'm not sure how class comparisons work exactly without looking it up, might be nice to comment here